core/domain/base_validators.py

```diff
@@ -0,0 +1,844 @@
+            item: ndb.Model. Entity to validate.
+        """
+        try:
+            model_domain_object_instance = (
+                cls._get_model_domain_object_instance(item))
+            if model_domain_object_instance is None:
+                # No domain object exists for this storage model class.
+                return
+            model_domain_object_instance.validate()
+        except Exception as e:
+            cls._add_error(
+                ERROR_CATEGORY_DOMAIN_OBJECT_CHECK,
+                'Entity id %s: Entity fails domain validation with the '
+                'error %s' % (item.id, e))
+
+    @classmethod
+    def _get_external_id_relationships(cls, item):
+        """Returns a mapping of external id to model class.
+
+        This should be implemented by subclasses.
+
+        Args:
+            item: ndb.Model. Entity to validate.
+
+        Returns:
+            list(ExternalModelFetcherDetails). A list whose values are
+            ExternalModelFetcherDetails instances each representing
+            the class and ids for a single type of external model to fetch.
+
+        Raises:
+            NotImplementedError. This function has not yet been implemented.
+        """
+        raise NotImplementedError(
+            'The _get_external_id_relationships() method is missing from the '
+            'derived class. It should be implemented in the derived class.')
+
+    @classmethod
+    def _validate_external_id_relationships(cls, item):
+        """Check whether the external id properties on the model correspond
+        to valid instances.
+
+        Args:
+            item: ndb.Model. Entity to validate.
+        """
+        for field_name, external_model_references in (
+                cls.field_name_to_external_model_references.items()):
+            for external_model_reference in external_model_references:
+                model = external_model_reference.model_instance
+
+                if model is None or model.deleted:
+                    model_class = external_model_reference.class_name
+                    model_id = external_model_reference.model_id
+                    cls._add_error(
+                        '%s %s' % (field_name, ERROR_CATEGORY_FIELD_CHECK),
+                        'Entity id %s: based on field %s having'
+                        ' value %s, expect model %s with id %s but it doesn\'t'
+                        ' exist' % (
+                            item.id, field_name, model_id,
+                            model_class.__name__, model_id))
+
+    @classmethod
+    def _fetch_field_name_to_external_model_references(cls, item):
+        """Fetch external models based on _get_external_id_relationships.
+
+        This should be called before we call other _validate methods.
+
+        Args:
+            item: ndb.Model. Entity to validate.
+        """
+        multiple_models_ids_to_fetch = {}
+
+        for external_model_fetcher in cls._get_external_id_relationships(item):
```

Comment:
`external_model_fetcher` variable sounds like a module name, mabe `external_model_details`?

Reply:
Updated to `external_model_fetcher_details`